Kubernetes 2-container pod staging.

jmchilton opened the pull request:
The tests included here are really very manual and brittle tests, but I'm not sure how to do much better without significantly more work. See documentation in the file docstring for more information.

Pulsar work on this (part of Pulsar 0.10.0) includes:
- https://github.com/galaxyproject/pulsar/pull/176
- https://github.com/galaxyproject/pulsar/pull/178

Fixes #4030. Implements first two big sections outlined in https://github.com/galaxyproject/galaxy/issues/7816.  Builds on #7854.

jmchilton committed:
Kubernetes 2-container pod staging.

The tests included here are really very manual and brittle tests, but I'm not sure how to do much better. See documentation in the file docstring for more information.

mvdbeek commented on test/integration/test_kubernetes_staging.py:
```diff
@@ -0,0 +1,117 @@
+"""Integration tests for Kubernetes pod staging.
+
+This case is way more brittle than the other integration tests distributed with Galaxy.
+This is because it requires Kubernetes, RabbitMQ, and the test itself needs to know what
```
Probably makes some sense for the test to spin up a rabbitmq docker container ?

jmchilton replied:
@mvdbeek Indeed, I started down that road but I got worried about something... whether the Kubernetes containers would be able to talk to the MQ container maybe... but that seems odd because you'd probably bind its ports to a port on the host. It is a good idea and would make this more robust I think.

@VJalili did you make any progress on MQ tests for Galaxy/Pulsar? If you're still working on or planning to work on that task, it might be worth shifting focus to this context and make this change here as @mvdbeek suggests. I'm not sure if this is a better start than what I had previously pointed you at or if it just more moving pieces to get right.

mvdbeek commented on test/unit/jobs/job_conf.sample_advanced.yml:
```diff
@@ -125,6 +125,13 @@ runners:
+    # Use a two-container Kubernetes to run jobs - one for staging and one
+    # for tool execution.
+    load: galaxy.jobs.runners.pulsar:PulsarKubernetesCoexecutionJobRunner
+    galaxy_url: http://docker.for.mac.localhost:8080
```
Since we figured that out for IEs as well, it seems that the recommendation for modern docker for mac (starting 18.03) is `host.docker.internal` (https://docs.docker.com/docker-for-mac/networking/)
```suggestion
    galaxy_url: http://host.docker.internal:8080
```

mvdbeek commented on test/integration/test_kubernetes_staging.py:
```diff
@@ -0,0 +1,117 @@
+the test host IP address will be relative to a container inside Kubernetes in order to
+communicate job status updates back.
+
+For this reason, this test will only work out of the box currently with Docker for Mac,
```
Surely not needed for this PR, but https://kind.sigs.k8s.io/ looks interesting for testing purposes.

mvdbeek commented on test/integration/test_kubernetes_staging.py:
```diff
@@ -0,0 +1,117 @@
+
+TOOL_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir, os.pardir, 'tools'))
+AMQP_URL = os.environ.get("GALAXY_TEST_AMQP_URL", "amqp://guest:guest@localhost:5672//")
+GALAXY_TEST_KUBERNETES_INFRASTRUCTURE_HOST = os.environ.get("GALAXY_TEST_KUBERNETES_INFRASTRUCTURE_HOST", "docker.for.mac.localhost")
```
```suggestion
GALAXY_TEST_KUBERNETES_INFRASTRUCTURE_HOST = os.environ.get("GALAXY_TEST_KUBERNETES_INFRASTRUCTURE_HOST", "host.docker.internal")
```

jmchilton committed:
Improvements to Kubernetes Pod Staging runner approach.

- Rev pulsar-galaxy-lib for improved error messages.
- Rev to a newer container - one that actually logs.
- More use of host.docker.internal to improve testing robustness.
- Add documentation for overriding container to use.

jmchilton commented:
Rebased and then updated with some improvements:

- Rev pulsar-galaxy-lib for improved error messages.
- Rev to a newer container - one that actually logs.
- More use of host.docker.internal to improve testing robustness.
- Add documentation for overriding container to use.

mvdbeek reviewed (approved):
Works like a charm! Can you just resolve the conflict ?

jmchilton commented:
Rebased with a conflict fix, thanks @mvdbeek. I have WIP on a generalization of this that just uses a single container with embedded Pulsar and normal Conda resolution if no container is configured. It might change the interface here a little but we are early in the release cycle and it would be good to get this merged before the GCC I think so I think this should still be merged. Thanks again!

mvdbeek commented:
Sounds good! I think it's actually quite nice to have this split, it'd potentially allow targeting the staging to a lower resource / cheaper target, right ?

jmchilton commented:
@mvdbeek Yeah... so there are really three options here:

- Stage + BioContainer as separate containers. (This PR)
- Stage, Dependency Resolution in same container. (Next PR)
- Stage + Dependency Resolution in separate containers.

I'm not sure that the third option would work with this WIP as is... but Pulsar certainly does have a nauseating number of knobs that can be tweaked and may be possible. We don't actually use init containers yet - so both containers would come online immediately and you'd be charged for the more expensive container the whole time. If we could do stage-in in an init-container and stage-out / biocontainer in separate containers though that might make sense to save staging in/out costs.

Update: xref init containers (https://kubernetes.io/docs/concepts/workloads/pods/init-containers/)

mvdbeek commented:
Yeah, that makes sense, I was guessing that was your idea after writing my comment ....

mvdbeek commented on lib/galaxy/jobs/runners/kubernetes.py:
```diff
@@ -4,7 +4,6 @@
 
 import logging
 import math
-import os
```
os is used again https://github.com/galaxyproject/galaxy/blob/5598305f07ddde4873a0943c9e5cea9a1f32f94f/lib/galaxy/jobs/runners/kubernetes.py#L161, which was added in the meantime.

VJalili commented:
@jmchilton could you please elaborate how data are staged?

jmchilton commented:
@VJalili 

Pulsar has many options for data staging - they can be configured on a per file basis - based on file type (input, output, config, reference, tool file, etc..) and/or based on regexes, globs, and prefixes. Each target specification has a corresponding transfer type - see https://pulsar.readthedocs.io/en/latest/galaxy_conf.html#data-staging.

The transfer types that expect a permanent Pulsar server (``transfer``) or that expect the target directory to exist after a Galaxy initiated setup (``copy``) will not work - but everything that is done by Pulsar should work:
- ``remote_transfer`` - Pulsar fetches files from the Galaxy job files API.
- ``remote_rsync_transfer`` - Pulsar uses rsync to copy the files. (untested in this context). 
- ``remote_copy`` - Pulsar can copy the file if K8S is somehow configured to have the target directory mounted.
- ``rewrite`` - just rewrite the paths - again assumes K8S is somehow configured to have the target directory mounted.

The default and easiest to use is ``remote_transfer`` though. This is what the test cases use.
